### PR TITLE
Dynamic provisioner for Cinder

### PR DESCRIPTION
**Pull request by pmorie**
Replaces #15606 

@thockin @saad-ali @jsafrane @markturansky @childsb

### CODE REVIEW

File: pkg/volume/cinder/cinder_util.go [hunk 1]
```diff
@@ -132,6 +132,40 @@ func (util *CinderDiskUtil) DetachDisk(cd *cinderVolumeCleaner) error {
 	return nil
 }
 
+func (util *CinderDiskUtil) DeleteVolume(cd *cinderVolumeDeleter) error {
+	cloud := cd.plugin.host.GetCloudProvider()
```
Comment (saad-ali) [inline pkg/volume/cinder/cinder_util.go]: Create a common util function `GetCiderCloudProvider(...)` to avoid code duplication.

File: pkg/volume/cinder/cinder_util.go [hunk 2]
```diff
@@ -132,6 +132,40 @@ func (util *CinderDiskUtil) DetachDisk(cd *cinderVolumeCleaner) error {
...
+func (util *CinderDiskUtil) CreateVolume(c *cinderVolumeProvisioner) (string, int, error) {
+	cloud := c.plugin.host.GetCloudProvider()
+	if cloud == nil {
+		glog.Errorf("Cloud provider not initialized properly")
+		return "", 0, errors.New("Cloud provider not initialized properly")
+	}
+
+	// convert to GB with rounding
+	const giga = 1024 * 1024 * 1024
+	volSize := int((c.options.Capacity.Value() + giga - 1) / giga)
```
Comment (saad-ali) [inline pkg/volume/cinder/cinder_util.go]: Same as AWS version, this deserves documentation. If there are implied units for options.Capacity that should be documented in the field's comments as well. Also, since the code is identical to the AWS version, factor it out to avoid code duplication.

Comment (saad-ali) [inline pkg/volume/cinder/cinder_util.go]: Looks like this is GB whereas the AWS is MB.

**Comment by saad-ali** [comment]
Same minor comments as the AWS version. Overall looks good.

File: pkg/volume/cinder/cinder_util.go (hunk 1 above)
Comment (jsafrane) [inline pkg/volume/cinder/cinder_util.go]: Done as separate patch in this PR, it modifies whole Cinder plugin, not only provisioning.

File: pkg/volume/cinder/cinder_util.go (hunk 2 above)
Comment (jsafrane) [inline pkg/volume/cinder/cinder_util.go]: Fixed in the same way as AWS. The code looks very similar to AWS, just with GB (AWS uses MB), but I think it will be quite different when we add QoS, as they are fundamentally different between various clouds.

**Comment by jsafrane** [comment]
@k8s-bot test this

**Comment by markturansky** [comment]
Unrelated flakes:

```
I1214 16:17:36.984470   23746 master.go:379] Node port range unspecified. Defaulting to 30000-32767.
[restful] 2015/12/14 16:17:37 log.go:30: [restful/swagger] listing is available at https://172.17.0.4:6443/swaggerapi/
[restful] 2015/12/14 16:17:37 log.go:30: [restful/swagger] https://172.17.0.4:6443/swaggerui/ is mapped to folder /swagger-ui/
F1214 16:17:37.122433   23746 controller.go:84] Unable to perform initial IP allocation check: unable to persist the updated service IP allocations: serviceipallocation "" cannot be updated: the provided resource version does not match
!!! [1214 16:17:49] Timed out waiting for apiserver:  to answer at http://127.0.0.1:8080/healthz; tried 25 waiting 0.5 between each
!!! Error in ./hack/test-update-storage-objects.sh:48
  'return 1' exited with status 1
Call stack:
  1: ./hack/test-update-storage-objects.sh:48 startApiServer(...)
  2: ./hack/test-update-storage-objects.sh:122 main(...)
Exiting with status 1
+++ [1214 16:17:49] Killing api server
```

**Comment by pmorie** [comment]
I alsao hit a flake in that test earlier.

File: pkg/volume/cinder/cinder_util.go (hunk 2 above)
Comment (saad-ali) [inline pkg/volume/cinder/cinder_util.go]: > The code looks very similar to AWS, just with GB (AWS uses MB), but I think it will be quite different when we add QoS, as they are fundamentally different between various clouds.

How so? As far as I can tell this just converts bytes to GiB (not GB) with some rounding. Feels like there should be some existing common util function to do this type of conversion, and if one doesn't exist it'd be nice to add one (and have this AWS and GCE code use it).

File: pkg/volume/cinder/cinder_util.go [hunk 3]
```diff
@@ -119,19 +117,51 @@ func (util *CinderDiskUtil) DetachDisk(cd *cinderVolumeCleaner) error {
...
+}
+
+func (util *CinderDiskUtil) CreateVolume(c *cinderVolumeProvisioner) (volumeID string, volumeSizeGB int, err error) {
+	cloud, err := c.plugin.getCloudProvider()
+	if err != nil {
+		return "", 0, err
+	}
+
+	volSizeBytes := c.options.Capacity.Value()
+	// Cinder works with gigabytes, convert to GB with rounding up
```
Comment (saad-ali) [inline pkg/volume/cinder/cinder_util.go]: GB or GiB?

Comment (jsafrane) [inline pkg/volume/cinder/cinder_util.go]: Comment fixed to GiB.

**Comment by jsafrane** [comment]
Build failure is expected, this PR depends on PR #18601.

**Comment by saad-ali** [comment]
LGTM. I will add LGTM label after #18601 is merged and this PR is rebased/squashed.

**Comment by saad-ali** [comment]
Manually merge to unblock Redhat. CC @jlowdermilk who's covering for scheduled on-call.
